File: src/Java/myservicebus-abstractions/src/main/java/com/myservicebus/MessageScheduler.java
package com.myservicebus;

import java.time.Duration;
import java.time.Instant;
import java.util.UUID;
import java.util.concurrent.CompletionStage;

import com.myservicebus.tasks.CancellationToken;

public interface MessageScheduler {
    SchedulingDurability getDurability();

    boolean supportsCancellation();

    <T> CompletionStage<ScheduledMessageHandle> schedulePublish(
            Instant scheduledTime,
            T message,
            CancellationToken cancellationToken);

    default <T> CompletionStage<ScheduledMessageHandle> schedulePublish(Instant scheduledTime, T message) {
        return schedulePublish(scheduledTime, message, CancellationToken.none());
    }

    <T> CompletionStage<ScheduledMessageHandle> schedulePublish(T message,
            Instant scheduledTime,
            CancellationToken cancellationToken);

    default <T> CompletionStage<ScheduledMessageHandle> schedulePublish(T message, Instant scheduledTime) {
        return schedulePublish(message, scheduledTime, CancellationToken.none());
    }

    default <T> CompletionStage<ScheduledMessageHandle> schedulePublish(T message, Duration delay,
            CancellationToken cancellationToken) {
        return schedulePublish(message, Instant.now().plus(delay), cancellationToken);
    }

    default <T> CompletionStage<ScheduledMessageHandle> schedulePublish(T message, Duration delay) {
        return schedulePublish(message, delay, CancellationToken.none());
    }

    <T> CompletionStage<ScheduledMessageHandle> scheduleSend(String destination,
            Instant scheduledTime,
            T message,
            CancellationToken cancellationToken);

    default <T> CompletionStage<ScheduledMessageHandle> scheduleSend(
            String destination,
            Instant scheduledTime,
            T message) {
        return scheduleSend(destination, scheduledTime, message, CancellationToken.none());
    }

    <T> CompletionStage<ScheduledMessageHandle> scheduleSend(String destination,
            T message,
            Instant scheduledTime,
            CancellationToken cancellationToken);

    default <T> CompletionStage<ScheduledMessageHandle> scheduleSend(String destination, T message, Instant scheduledTime) {
        return scheduleSend(destination, message, scheduledTime, CancellationToken.none());
    }

    default <T> CompletionStage<ScheduledMessageHandle> scheduleSend(String destination, T message, Duration delay,
            CancellationToken cancellationToken) {
        return scheduleSend(destination, message, Instant.now().plus(delay), cancellationToken);
    }

    default <T> CompletionStage<ScheduledMessageHandle> scheduleSend(String destination, T message, Duration delay) {
        return scheduleSend(destination, message, delay, CancellationToken.none());
    }

    CompletionStage<ScheduleCancellationResult> cancelScheduledPublish(UUID tokenId, CancellationToken cancellationToken);

    default CompletionStage<ScheduleCancellationResult> cancelScheduledPublish(UUID tokenId) {
        return cancelScheduledPublish(tokenId, CancellationToken.none());
    }

    default CompletionStage<ScheduleCancellationResult> cancelScheduledPublish(ScheduledMessageHandle handle, CancellationToken cancellationToken) {
        return cancelScheduledPublish(handle.getTokenId(), cancellationToken);
    }

    default CompletionStage<ScheduleCancellationResult> cancelScheduledPublish(ScheduledMessageHandle handle) {
        return cancelScheduledPublish(handle, CancellationToken.none());
    }

    CompletionStage<ScheduleCancellationResult> cancelScheduledSend(UUID tokenId, CancellationToken cancellationToken);

    default CompletionStage<ScheduleCancellationResult> cancelScheduledSend(UUID tokenId) {
        return cancelScheduledSend(tokenId, CancellationToken.none());
    }

    default CompletionStage<ScheduleCancellationResult> cancelScheduledSend(ScheduledMessageHandle handle, CancellationToken cancellationToken) {
        return cancelScheduledSend(handle.getTokenId(), cancellationToken);
    }

    default CompletionStage<ScheduleCancellationResult> cancelScheduledSend(ScheduledMessageHandle handle) {
        return cancelScheduledSend(handle, CancellationToken.none());
    }
}

File: src/MyServiceBus/MessageScheduler.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace MyServiceBus;

public class MessageScheduler : IMessageScheduler
{
    private readonly IScheduleMessageProvider provider;

    public MessageScheduler(IScheduleMessageProvider provider)
    {
        this.provider = provider;
    }

    public ScheduleMessageProviderDurability Durability => provider.Durability;

    public bool SupportsCancellation => provider.SupportsCancellation;

    public Task<ScheduledMessageHandle> SchedulePublish<T>(DateTime scheduledTime, T message, CancellationToken cancellationToken = default) where T : class
        => provider.SchedulePublish(scheduledTime, message, cancellationToken);

    public Task<ScheduledMessageHandle> SchedulePublish<T>(T message, DateTime scheduledTime, CancellationToken cancellationToken = default) where T : class
        => SchedulePublish(scheduledTime, message, cancellationToken);

    public Task<ScheduledMessageHandle> SchedulePublish<T>(T message, TimeSpan delay, CancellationToken cancellationToken = default) where T : class
        => SchedulePublish(DateTime.UtcNow + delay, message, cancellationToken);

    public Task<ScheduledMessageHandle> ScheduleSend<T>(Uri destination, DateTime scheduledTime, T message, CancellationToken cancellationToken = default) where T : class
        => provider.ScheduleSend(destination, scheduledTime, message, cancellationToken);

    public Task<ScheduledMessageHandle> ScheduleSend<T>(Uri destination, T message, DateTime scheduledTime, CancellationToken cancellationToken = default) where T : class
        => ScheduleSend(destination, scheduledTime, message, cancellationToken);

    public Task<ScheduledMessageHandle> ScheduleSend<T>(Uri destination, T message, TimeSpan delay, CancellationToken cancellationToken = default) where T : class
        => ScheduleSend(destination, DateTime.UtcNow + delay, message, cancellationToken);

    public Task CancelScheduledPublish(Guid tokenId, CancellationToken cancellationToken = default)
        => provider.Cancel(tokenId, cancellationToken);

    public Task CancelScheduledPublish(ScheduledMessageHandle handle, CancellationToken cancellationToken = default)
        => CancelScheduledPublish(handle.TokenId, cancellationToken);

    public Task CancelScheduledSend(Guid tokenId, CancellationToken cancellationToken = default)
        => provider.Cancel(tokenId, cancellationToken);

    public Task CancelScheduledSend(ScheduledMessageHandle handle, CancellationToken cancellationToken = default)
        => CancelScheduledSend(handle.TokenId, cancellationToken);
}
